File: python/adobe/core/dom.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from .session import HostSession

_DOM_REFERENCE_KEY = "$adobepyRef"
_DOM_TYPE_KEY = "$adobepyType"


class DomObject:
    """Opaque reference to an object owned by an Adobe host bridge."""

    def __init__(self, namespace: "DomNamespace", reference: str, type_name: str | None = None) -> None:
        self._namespace = namespace
        self.reference = reference
        self.type_name = type_name

    def get(self, member: str | int, *, timeout_ms: int | None = None) -> Any:
        return self._namespace.get(self, member, timeout_ms=timeout_ms)

    async def get_async(self, member: str | int, *, timeout_ms: int | None = None) -> Any:
        return await self._namespace.get_async(self, member, timeout_ms=timeout_ms)

    def set(
        self,
        member: str | int,
        value: Any,
        *,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> Any:
        return self._namespace.set(self, member, value, command_name=command_name, timeout_ms=timeout_ms)

    async def set_async(
        self,
        member: str | int,
        value: Any,
        *,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> Any:
        return await self._namespace.set_async(
            self,
            member,
            value,
            command_name=command_name,
            timeout_ms=timeout_ms,
        )

    def call(
        self,
        member: str | int,
        *args: Any,
        command_name: str | None = None,
        mutating: bool = False,
        timeout_ms: int | None = None,
    ) -> Any:
        return self._namespace.call(
            self,
            member,
            *args,
            command_name=command_name,
            mutating=mutating,
            timeout_ms=timeout_ms,
        )

    async def call_async(
        self,
        member: str | int,
        *args: Any,
        command_name: str | None = None,
        mutating: bool = False,
        timeout_ms: int | None = None,
    ) -> Any:
        return await self._namespace.call_async(
            self,
            member,
            *args,
            command_name=command_name,
            mutating=mutating,
            timeout_ms=timeout_ms,
        )

    def construct(
        self,
        member: str,
        *args: Any,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> "DomObject":
        return self._namespace.construct(
            self,
            member,
            *args,
            command_name=command_name,
            timeout_ms=timeout_ms,
        )

    async def construct_async(
        self,
        member: str,
        *args: Any,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> "DomObject":
        return await self._namespace.construct_async(
            self,
            member,
            *args,
            command_name=command_name,
            timeout_ms=timeout_ms,
        )

    def keys(self, *, timeout_ms: int | None = None) -> list[str]:
        return self._namespace.keys(self, timeout_ms=timeout_ms)

    async def keys_async(self, *, timeout_ms: int | None = None) -> list[str]:
        return await self._namespace.keys_async(self, timeout_ms=timeout_ms)

    def snapshot(self, *members: str | int, timeout_ms: int | None = None) -> dict[str, Any]:
        return self._namespace.snapshot(self, *members, timeout_ms=timeout_ms)

    async def snapshot_async(self, *members: str | int, timeout_ms: int | None = None) -> dict[str, Any]:
        return await self._namespace.snapshot_async(self, *members, timeout_ms=timeout_ms)

    def release(self, *, timeout_ms: int | None = None) -> bool:
        return self._namespace.release(self, timeout_ms=timeout_ms)

    async def release_async(self, *, timeout_ms: int | None = None) -> bool:
        return await self._namespace.release_async(self, timeout_ms=timeout_ms)

    def _wire_value(self) -> dict[str, str]:
        value = {_DOM_REFERENCE_KEY: self.reference}
        if self.type_name:
            value[_DOM_TYPE_KEY] = self.type_name
        return value

    def __repr__(self) -> str:
        suffix = f" type={self.type_name!r}" if self.type_name else ""
        return f"DomObject(reference={self.reference!r}{suffix})"


class DomNamespace:
    """Structured, non-eval access to a host's complete official object model."""

    def __init__(self, session: "HostSession") -> None:
        self._session = session

    def root(self, name: str = "app", *, timeout_ms: int | None = None) -> DomObject:
        result = self._invoke("root", name, timeout_ms=timeout_ms)
        return _require_dom_object(result, "dom.root")

    async def root_async(self, name: str = "app", *, timeout_ms: int | None = None) -> DomObject:
        result = await self._invoke_async("root", name, timeout_ms=timeout_ms)
        return _require_dom_object(result, "dom.root")

    def get(self, receiver: DomObject, member: str | int, *, timeout_ms: int | None = None) -> Any:
        return self._invoke("get", receiver, member, timeout_ms=timeout_ms)

    async def get_async(self, receiver: DomObject, member: str | int, *, timeout_ms: int | None = None) -> Any:
        return await self._invoke_async("get", receiver, member, timeout_ms=timeout_ms)

    def set(
        self,
        receiver: DomObject,
        member: str | int,
        value: Any,
        *,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> Any:
        return self._invoke(
            "set",
            receiver,
            member,
            value,
            command_name=command_name,
            mutating=True,
            timeout_ms=timeout_ms,
        )

    async def set_async(
        self,
        receiver: DomObject,
        member: str | int,
        value: Any,
        *,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> Any:
        return await self._invoke_async(
            "set",
            receiver,
            member,
            value,
            command_name=command_name,
            mutating=True,
            timeout_ms=timeout_ms,
        )

    def call(
        self,
        receiver: DomObject,
        member: str | int,
        *args: Any,
        command_name: str | None = None,
        mutating: bool = False,
        timeout_ms: int | None = None,
    ) -> Any:
        return self._invoke(
            "call",
            receiver,
            member,
            list(args),
            command_name=command_name,
            mutating=mutating,
            timeout_ms=timeout_ms,
        )

    async def call_async(
        self,
        receiver: DomObject,
        member: str | int,
        *args: Any,
        command_name: str | None = None,
        mutating: bool = False,
        timeout_ms: int | None = None,
    ) -> Any:
        return await self._invoke_async(
            "call",
            receiver,
            member,
            list(args),
            command_name=command_name,
            mutating=mutating,
            timeout_ms=timeout_ms,
        )

    def construct(
        self,
        receiver: DomObject,
        member: str,
        *args: Any,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> DomObject:
        result = self._invoke(
            "construct",
            receiver,
            member,
            list(args),
            command_name=command_name,
            mutating=True,
            timeout_ms=timeout_ms,
        )
        return _require_dom_object(result, "dom.construct")

    async def construct_async(
        self,
        receiver: DomObject,
        member: str,
        *args: Any,
        command_name: str | None = None,
        timeout_ms: int | None = None,
    ) -> DomObject:
        result = await self._invoke_async(
            "construct",
            receiver,
            member,
            list(args),
            command_name=command_name,
            mutating=True,
            timeout_ms=timeout_ms,
        )
        return _require_dom_object(result, "dom.construct")

    def keys(self, receiver: DomObject, *, timeout_ms: int | None = None) -> list[str]:
        result = self._invoke("keys", receiver, timeout_ms=timeout_ms)
        return [str(item) for item in result]

    async def keys_async(self, receiver: DomObject, *, timeout_ms: int | None = None) -> list[str]:
        result = await self._invoke_async("keys", receiver, timeout_ms=timeout_ms)
        return [str(item) for item in result]

    def snapshot(
        self,
        receiver: DomObject,
        *members: str | int,
        timeout_ms: int | None = None,
    ) -> dict[str, Any]:
        result = self._invoke("snapshot", receiver, list(members) if members else None, timeout_ms=timeout_ms)
        if not isinstance(result, dict):
            raise TypeError("dom.snapshot returned a non-object result")
        return result

    async def snapshot_async(
        self,
        receiver: DomObject,
        *members: str | int,
        timeout_ms: int | None = None,
    ) -> dict[str, Any]:
        result = await self._invoke_async(
            "snapshot",
            receiver,
            list(members) if members else None,
            timeout_ms=timeout_ms,
        )
        if not isinstance(result, dict):
            raise TypeError("dom.snapshot returned a non-object result")
        return result

    def release(self, receiver: DomObject, *, timeout_ms: int | None = None) -> bool:
        return bool(self._invoke("release", receiver, timeout_ms=timeout_ms))

    async def release_async(self, receiver: DomObject, *, timeout_ms: int | None = None) -> bool:
        return bool(await self._invoke_async("release", receiver, timeout_ms=timeout_ms))

    def _invoke(
        self,
        method: str,
        *args: Any,
        command_name: str | None = None,
        mutating: bool = False,
        timeout_ms: int | None = None,
    ) -> Any:
        result = self._session.invoke(
            "dom",
            method,
            *(_encode_dom_value(arg, self) for arg in args),
            options=_dom_options(command_name, mutating, timeout_ms),
        )
        return self._decode(result)

    async def _invoke_async(
        self,
        method: str,
        *args: Any,
        command_name: str | None = None,
        mutating: bool = False,
        timeout_ms: int | None = None,
    ) -> Any:
        result = await self._session.invoke_async(
            "dom",
            method,
            *(_encode_dom_value(arg, self) for arg in args),
            options=_dom_options(command_name, mutating, timeout_ms),
        )
        return self._decode(result)

    def _decode(self, value: Any) -> Any:
        if isinstance(value, list):
            return [self._decode(item) for item in value]
        if isinstance(value, dict):
            reference = value.get(_DOM_REFERENCE_KEY)
            if isinstance(reference, str):
                type_name = value.get(_DOM_TYPE_KEY)
                return DomObject(self, reference, str(type_name) if type_name is not None else None)
            return {key: self._decode(item) for key, item in value.items()}
        return value


def _encode_dom_value(value: Any, namespace: DomNamespace) -> Any:
    if isinstance(value, DomObject):
        if value._namespace is not namespace:
            raise ValueError("DomObject belongs to a different host session")
        return value._wire_value()
    if isinstance(value, list):
        return [_encode_dom_value(item, namespace) for item in value]
    if isinstance(value, tuple):
        return [_encode_dom_value(item, namespace) for item in value]
    if isinstance(value, dict):
        return {key: _encode_dom_value(item, namespace) for key, item in value.items()}
    return value


def _dom_options(command_name: str | None, mutating: bool, timeout_ms: int | None) -> dict[str, Any]:
    options: dict[str, Any] = {}
    if timeout_ms is not None:
        options["timeoutMs"] = timeout_ms
    if command_name is not None:
        options["commandName"] = command_name
    if mutating:
        options["modal"] = True
    return options


def _require_dom_object(value: Any, operation: str) -> DomObject:
    if not isinstance(value, DomObject):
        raise TypeError(f"{operation} returned a non-reference result")
    return value

File: bridges/uxp/core/src/dom.ts
import { BridgeRpcError, ERROR_HOST_SCRIPT, methodNotFound } from "./errors";
import type { RpcRequest } from "./protocol";
import { asString, isObject, maybePromise, property } from "./runtime";

const REFERENCE_KEY = "$adobepyRef";
const TYPE_KEY = "$adobepyType";
const BLOCKED_MEMBERS = new Set(["__proto__", "constructor", "prototype"]);

type DomRootResolver = (name: string) => unknown;
type DomMutationRunner = (
  request: RpcRequest,
  defaultCommandName: string,
  operation: () => Promise<unknown>,
) => Promise<unknown>;

export interface DomRuntimeOptions {
  roots: DomRootResolver;
  runMutation?: DomMutationRunner;
}

/**
 * Structured access to an official UXP host object model.
 *
 * The runtime deliberately exchanges opaque object references instead of
 * evaluating caller-provided JavaScript. This keeps the complete, versioned
 * Adobe DOM reachable while preserving JSON-RPC, authentication, target, and
 * command-boundary contracts.
 */
export class DomRuntime {
  private readonly references = new Map<string, unknown>();
  private readonly reverseReferences = new WeakMap<object, string>();
  private nextReference = 1;

  constructor(private readonly options: DomRuntimeOptions) {}

  async dispatch(request: RpcRequest): Promise<unknown> {
    if (request.namespace !== "dom") {
      throw new BridgeRpcError(ERROR_HOST_SCRIPT, `invalid DOM namespace ${request.namespace}`);
    }

    if (request.method === "root") {
      const name = requiredString(request.args?.[0], "root name");
      const value = this.options.roots(name);
      if (value === undefined || value === null) {
        throw new BridgeRpcError(ERROR_HOST_SCRIPT, `official DOM root '${name}' is unavailable`);
      }
      return this.encode(value);
    }

    if (request.method === "get") {
      const receiver = this.resolve(request.args?.[0]);
      const member = requiredMember(request.args?.[1]);
      return this.encode(await maybePromise(readMember(receiver, member)));
    }

    if (request.method === "set") {
      return this.runMutation(request, "Set official DOM property", async () => {
        const receiver = this.resolve(request.args?.[0]);
        const member = requiredMember(request.args?.[1]);
        const value = this.decode(request.args?.[2]);
        writeMember(receiver, member, value);
        return this.encode(await maybePromise(readMember(receiver, member)));
      });
    }

    if (request.method === "call") {
      const operation = async () => {
        const receiver = this.resolve(request.args?.[0]);
        const member = requiredMember(request.args?.[1]);
        const callable = readMember(receiver, member);
        if (typeof callable !== "function") {
          throw new BridgeRpcError(ERROR_HOST_SCRIPT, `official DOM member '${String(member)}' is not callable`);
        }
        const args = this.decodeArgs(request.args?.[2]);
        return this.encode(await maybePromise(callable.apply(receiver, args)));
      };
      if (request.options?.modal === true || typeof request.options?.commandName === "string") {
        return this.runMutation(request, `Call official DOM method ${String(request.args?.[1] ?? "")}`, operation);
      }
      return operation();
    }

    if (request.method === "construct") {
      return this.runMutation(request, "Construct official DOM object", async () => {
        const receiver = this.resolve(request.args?.[0]);
        const member = requiredMember(request.args?.[1]);
        const constructor = readMember(receiver, member);
        if (typeof constructor !== "function") {
          throw new BridgeRpcError(ERROR_HOST_SCRIPT, `official DOM member '${String(member)}' is not a constructor`);
        }
        return this.encode(Reflect.construct(constructor, this.decodeArgs(request.args?.[2])));
      });
    }

    if (request.method === "keys") {
      return collectKeys(this.resolve(request.args?.[0]));
    }

    if (request.method === "snapshot") {
      const receiver = this.resolve(request.args?.[0]);
      const requested = request.args?.[1];
      const members = Array.isArray(requested)
        ? requested.map(requiredMember)
        : collectSnapshotKeys(receiver);
      const snapshot: Record<string, unknown> = {};
      for (const member of members) {
        try {
          const value = await maybePromise(readMember(receiver, member));
          if (typeof value !== "function") snapshot[String(member)] = this.encode(value);
        } catch (error) {
          snapshot[String(member)] = {
            $adobepyError: error instanceof Error ? error.message : String(error),
          };
        }
      }
      return snapshot;
    }

    if (request.method === "release") {
      const reference = referenceId(request.args?.[0]);
      const value = this.references.get(reference);
      if ((typeof value === "object" && value !== null) || typeof value === "function") {
        this.reverseReferences.delete(value as object);
      }
      return this.references.delete(reference);
    }

    methodNotFound("dom", request.method);
  }

  private async runMutation(
    request: RpcRequest,
    defaultCommandName: string,
    operation: () => Promise<unknown>,
  ): Promise<unknown> {
    if (this.options.runMutation) {
      return this.options.runMutation(request, defaultCommandName, operation);
    }
    return operation();
  }

  private encode(value: unknown): unknown {
    if (value === undefined || value === null) return null;
    if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") return value;
    if (typeof value === "bigint") return value.toString();
    if (typeof value === "symbol") return value.description ?? value.toString();
    if (Array.isArray(value)) return value.map((item) => this.encode(item));

    const objectValue = value as object;
    const existing = this.reverseReferences.get(objectValue);
    if (existing) {
      return {
        [REFERENCE_KEY]: existing,
        [TYPE_KEY]: domTypeName(value),
      };
    }

    const reference = `uxp_${this.nextReference++}`;
    this.references.set(reference, value);
    this.reverseReferences.set(objectValue, reference);
    return {
      [REFERENCE_KEY]: reference,
      [TYPE_KEY]: domTypeName(value),
    };
  }

  private decode(value: unknown): unknown {
    if (Array.isArray(value)) return value.map((item) => this.decode(item));
    if (!isObject(value)) return value;
    if (typeof value[REFERENCE_KEY] === "string") return this.resolve(value);
    return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, this.decode(item)]));
  }

  private decodeArgs(value: unknown): unknown[] {
    if (value === undefined) return [];
    if (!Array.isArray(value)) {
      throw new BridgeRpcError(ERROR_HOST_SCRIPT, "official DOM call arguments must be an array");
    }
    return value.map((item) => this.decode(item));
  }

  private resolve(value: unknown): unknown {
    const reference = referenceId(value);
    if (!this.references.has(reference)) {
      throw new BridgeRpcError(ERROR_HOST_SCRIPT, `official DOM reference '${reference}' is stale or unknown`);
    }
    return this.references.get(reference);
  }
}

function referenceId(value: unknown): string {
  const reference = isObject(value) ? value[REFERENCE_KEY] : undefined;
  if (typeof reference !== "string" || reference.length === 0) {
    throw new BridgeRpcError(ERROR_HOST_SCRIPT, `expected an object containing '${REFERENCE_KEY}'`);
  }
  return reference;
}

function requiredString(value: unknown, label: string): string {
  const result = asString(value);
  if (!result) throw new BridgeRpcError(ERROR_HOST_SCRIPT, `${label} is required`);
  return result;
}

function requiredMember(value: unknown): string | number {
  if (typeof value !== "string" && typeof value !== "number") {
    throw new BridgeRpcError(ERROR_HOST_SCRIPT, "official DOM member must be a string or array index");
  }
  if (typeof value === "string" && BLOCKED_MEMBERS.has(value)) {
    throw new BridgeRpcError(ERROR_HOST_SCRIPT, `official DOM member '${value}' is not accessible`);
  }
  return value;
}

function readMember(receiver: unknown, member: string | number): unknown {
  if ((typeof receiver !== "object" && typeof receiver !== "function") || receiver === null) {
    throw new BridgeRpcError(ERROR_HOST_SCRIPT, `cannot read '${String(member)}' from a primitive value`);
  }
  return (receiver as Record<string | number, unknown>)[member];
}

function writeMember(receiver: unknown, member: string | number, value: unknown): void {
  if ((typeof receiver !== "object" && typeof receiver !== "function") || receiver === null) {
    throw new BridgeRpcError(ERROR_HOST_SCRIPT, `cannot write '${String(member)}' on a primitive value`);
  }
  (receiver as Record<string | number, unknown>)[member] = value;
}

function collectKeys(value: unknown): string[] {
  if ((typeof value !== "object" && typeof value !== "function") || value === null) return [];
  const keys = new Set<string>();
  let current: object | null = value as object;
  for (let depth = 0; current && depth < 8; depth += 1) {
    try {
      for (const key of Object.getOwnPropertyNames(current)) {
        if (!BLOCKED_MEMBERS.has(key) && key !== "constructor") keys.add(key);
      }
      current = Object.getPrototypeOf(current);
    } catch {
      break;
    }
  }
  return Array.from(keys).sort();
}

function collectSnapshotKeys(value: unknown): string[] {
  if ((typeof value !== "object" && typeof value !== "function") || value === null) return [];
  try {
    return Object.keys(value).filter((key) => !BLOCKED_MEMBERS.has(key)).sort();
  } catch {
    return [];
  }
}

function domTypeName(value: unknown): string {
  if (typeof value === "function") return value.name || "Function";
  if (!isObject(value)) return typeof value;
  const typename = asString(property(value, "typename")) ?? asString(property(value, "typeName"));
  if (typename) return typename;
  const constructor = property<{ name?: unknown }>(value, "constructor");
  return asString(constructor?.name) ?? "Object";
}
